### PR TITLE
Add service layer for the letters count report

### PR DESCRIPTION
**Pull request by aliveni**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/BPS-626

### Change description ###
Added service layer for the letters count report.
The Reports Service excludes the test service, adds a row with count "0" when no letters are uplaoded for the services.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java [hunk 1]
```diff
@@ -1,16 +1,63 @@
...
+        ZeroRowFiller zeroRowFiller
+    ) {
+        this.repo = repo;
+        this.serviceFolderMapping = serviceFolderMapping;
+        this.zeroRowFiller = zeroRowFiller;
+    }
+
     public List<LettersCountSummary> getCountFor(LocalDate date) {
-        throw new NotImplementedException("Not yet implemented");
+        LocalDateTime dateTimeFrom = formatDateTime(date.minusDays(1), LocalTime.of(17, 0, 0));
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: Can we make this configurable ? I mean FTP downtime 16.00-17.00

Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: Sure. we already have the configuration for  FTP downtime. I'll re-use them

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java [hunk 2]
```diff
@@ -1,16 +1,63 @@
...
+    }
+
+    private LettersCountSummary fromDb(ServiceLettersCountSummary dbSummary) {
+        return new LettersCountSummary(
+            serviceFolderMapping.getFolderFor(dbSummary.getService()).orElse(null),
+            dbSummary.getUploaded());
+    }
+
+    private LocalDateTime formatDateTime(LocalDate date, LocalTime time) {
+        ZonedDateTime zonedDateTime = ZonedDateTime.of(date, time, ZoneId.of(EUROPE_LONDON));
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: 👍

File: src/test/java/uk/gov/hmcts/reform/sendletter/services/ReportsServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+        //then
+        assertThat(result).isNotEmpty()
+            .hasSize(1)
+            .usingFieldByFieldElementComparator()
+            .containsExactlyInAnyOrder(new LettersCountSummary("ServiceA", 10));
+    }
+
+    @Test
+    void should_return_empty_list_for_the_date_when_repo_returns_empty_collection() {
```
Comment (nitinprabhu) [inline src/test/java/uk/gov/hmcts/reform/sendletter/services/ReportsServiceTest.java]: should it be not we return 0 when we don't have any records for a service ?

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ZeroRowFiller.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+
+@Component
+public class ZeroRowFiller {
+
+    private final List<String> serviceFolders;
+
+    private final ServiceFolderMapping serviceFolderMapping;
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/ZeroRowFiller.java]: not used

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ZeroRowFiller.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+    public ZeroRowFiller(ServiceFolderMapping serviceFolderMapping) {
+        this.serviceFolderMapping = serviceFolderMapping;
+        this.serviceFolders = Lists.newArrayList(serviceFolderMapping.getFolders());
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/ZeroRowFiller.java]: I think we need service names, not folder names

File: src/test/java/uk/gov/hmcts/reform/sendletter/services/ReportsServiceTest.java (hunk 3 above)
Comment (aliveni) [reply src/test/java/uk/gov/hmcts/reform/sendletter/services/ReportsServiceTest.java]: yeah it returns 0 when no letters uploaded for the service. I'll change the test name

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java [hunk 6]
```diff
@@ -1,16 +1,63 @@
...
+        String formattedDateTime = zonedDateTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss"));
+        return LocalDateTime.parse(formattedDateTime);
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: would `zonedDateTime.toLocalDateTime()` be enough here?

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java [hunk 7]
```diff
@@ -1,16 +1,72 @@
...
+import java.time.format.DateTimeFormatter;
 import java.util.List;
 
+import static java.util.stream.Collectors.toList;
+import static uk.gov.hmcts.reform.sendletter.util.TimeZones.EUROPE_LONDON;
+
 @Service
 public class ReportsService {
 
+    private static final String TEST_SERVICE = "BULKPRINT";
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: it's `send_letter_tests`

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java [hunk 8]
```diff
@@ -1,16 +1,72 @@
...
+        return zeroRowFiller.fill(
+            repo.countByDate(dateTimeFrom, dateTimeTo).stream().map(this::fromDb).collect(toList()))
+            .stream()
+            .filter(summary -> !summary.service.equals(TEST_SERVICE))
+            .collect(toList()); //exclude test service
+    }
+
+    private LettersCountSummary fromDb(ServiceLettersCountSummary dbSummary) {
+        return new LettersCountSummary(
+            serviceFolderMapping.getFolderFor(dbSummary.getService()).orElse(null),
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: do we want nulls in the outcome too?

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java [hunk 9]
```diff
@@ -1,16 +1,72 @@
...
+        this.repo = repo;
+        this.serviceFolderMapping = serviceFolderMapping;
+        this.zeroRowFiller = zeroRowFiller;
+        this.downtimeFromHour = downtimeFromHour;
+        this.downtimeToHour = downtimeToHour;
+    }
+
     public List<LettersCountSummary> getCountFor(LocalDate date) {
-        throw new NotImplementedException("Not yet implemented");
+        LocalDateTime dateTimeFrom = formatDateTime(date.minusDays(1), LocalTime.parse(downtimeFromHour));
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: should be `downtimeToHour` (and vice versa below)

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java [hunk 10]
```diff
@@ -1,16 +1,72 @@
...
+        LocalDateTime dateTimeTo = formatDateTime(date, LocalTime.parse(downtimeToHour));
+
+        return zeroRowFiller.fill(
+            repo.countByDate(dateTimeFrom, dateTimeTo).stream().map(this::fromDb).collect(toList()))
+            .stream()
+            .filter(summary -> !summary.service.equals(TEST_SERVICE))
+            .collect(toList()); //exclude test service
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: wrong line for comment

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java (hunk 7 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: I'm using the folder names in the report

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java [hunk 11]
```diff
@@ -1,16 +1,72 @@
...
+        this.downtimeToHour = downtimeToHour;
+    }
+
     public List<LettersCountSummary> getCountFor(LocalDate date) {
-        throw new NotImplementedException("Not yet implemented");
+        LocalDateTime dateTimeFrom = formatDateTime(date.minusDays(1), LocalTime.parse(downtimeFromHour));
+        LocalDateTime dateTimeTo = formatDateTime(date, LocalTime.parse(downtimeToHour));
+
+        return zeroRowFiller.fill(
+            repo.countByDate(dateTimeFrom, dateTimeTo).stream().map(this::fromDb).collect(toList()))
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: something for different pr: repo can return stream as well. just need to remember to close it upon direct usage

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java (hunk 8 above)
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: should be service name, not folder

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java (hunk 7 above)
Comment (nitinprabhu) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: that's correct I think else you would have `divorce_frontend` for divorce

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java (hunk 8 above)
Comment (nitinprabhu) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: we will have `divorce_frontend` if we use service name and I am sure we would expect to see `DIVORCE` rather then `divorce_frontend`

Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: the service names are "cmc_claim_store", "divorce_frontend", "probate_backend" and "sscs". 
The folder names would be clearer on the report which would be CMC, DIVORCE, PROBATE and SSCS

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java (hunk 7 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: The model suggests it's a service name @aliveni 
```
public class LettersCountSummary {

    public final String service;
    public final int uploaded;

    public LettersCountSummary(String service, int uploaded) {
        this.service = service;
        this.uploaded = uploaded;
    }
}
```
I assume business is interested in service names, not folders on FTP server.

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java (hunk 8 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: good thing it's not `TO_XEROX` anymore

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: I still think it would be more sensible to use service name instead of name of a folder on some FTP server.
Folder names just happen to looks like something that business would like.
If service names are not clear (to be checked) we could add a mapping between actual service name and service reports display name.

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java (hunk 6 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: wouldn't
```
LocalDateTime dt = LocalDateTime.of(datePart, timePart);
```
work?

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java (hunk 2 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: I think dates is db are UTC

File: src/test/java/uk/gov/hmcts/reform/sendletter/services/ReportsServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+        LocalTime timeTo = LocalTime.of(16, 0, 0);
+
+        //given
+        given(repository.countByDate(formatDateTime(date.minusDays(1), timeFrom), formatDateTime(date, timeTo)))
+            .willReturn(asList(
+                new ServiceLettersCount("aService", 10),
+                new ServiceLettersCount("bService", 20)
+            ));
+
+        given(serviceFolderMapping.getFolderFor("aService")).willReturn(Optional.of("ServiceA"));
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/sendletter/services/ReportsServiceTest.java]: did you consider `folderA` and `serviceA`
instead of `aService` and `ServiceA`?

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java (hunk 6 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: It would work but it won't be with Timezone

File: src/test/java/uk/gov/hmcts/reform/sendletter/services/ReportsServiceTest.java (hunk 3 above)
Comment (rkondratowicz) [reply src/test/java/uk/gov/hmcts/reform/sendletter/services/ReportsServiceTest.java]: @nitinprabhu `ZeroRowFiller` adds rows with zeros, which in this test is mocked to not do anything

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java (hunk 8 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: something would be reaaaally wrong if there's no folder for a service.
Will not be a problem if we simply return a service name though.

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java (hunk 2 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: updating the code to use UTC format

File: src/test/java/uk/gov/hmcts/reform/sendletter/services/ReportsServiceTest.java (hunk 3 above)
Comment (aliveni) [reply src/test/java/uk/gov/hmcts/reform/sendletter/services/ReportsServiceTest.java]: yeah, I just renamed the test

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ZeroRowFiller.java (hunk 4 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ZeroRowFiller.java]: removed

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java (hunk 6 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: updated to use UTC format

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java (hunk 8 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: we had a team discussion and decided to create a ticket to add service name property to the mappings config.

Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: @doncem As discussed, updating to filter the null values (just in case we have nulls values)

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java (hunk 11 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: I'll do it in a separate pr as it involves repository changes

File: src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java (hunk 7 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/sendletter/services/ReportsService.java]: I'm leaving the model as it is for now as we will new config for the service names

File: src/test/java/uk/gov/hmcts/reform/sendletter/services/ReportsServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+        List<LettersCountSummary> result = service.getCountFor(date);
+
+        //then
+        assertThat(result).isEmpty();
+    }
+
+    private LocalDateTime formatDateTime(LocalDate date, LocalTime time) {
+        ZonedDateTime zonedDateTime = ZonedDateTime.of(date, time, ZoneId.from(ZoneOffset.UTC));
+        return zonedDateTime.toLocalDateTime();
+    }
```
Comment (doncem) [inline src/test/java/uk/gov/hmcts/reform/sendletter/services/ReportsServiceTest.java]: I think this is second usage of same implementation. Perhaps worth moving to `uk.gov.hmcts.reform.sendletter.util.TimeZones` utility class? Happy for it to go in another PR

File: src/test/java/uk/gov/hmcts/reform/sendletter/services/ReportsServiceTest.java (hunk 12 above)
Comment (aliveni) [reply src/test/java/uk/gov/hmcts/reform/sendletter/services/ReportsServiceTest.java]: updated the tests

File: src/test/java/uk/gov/hmcts/reform/sendletter/services/ReportsServiceTest.java (hunk 13 above)
Comment (aliveni) [reply src/test/java/uk/gov/hmcts/reform/sendletter/services/ReportsServiceTest.java]: moved it to TimeZones utility class
